### PR TITLE
Multiple port support

### PR DESCRIPTION
**Pull request by rosmo**
Added support for services exposing multiple ports to use multiple NEGs, which can be used with the new annotation name and format. Old annotation format is still supported, but deprecated.

Updated documentation to match.

### CODE REVIEW

**Comment by rosmo** [comment]
I've added now support for multiple backends per port and support for regional backend services (eg. L7 ILB). This means you can attach a service both to an external load balancer and to an internal one. Appropriate tests were added as well.

**Comment by rwkarg** [comment]
Regional (L7 ILB) support will address #23

**Comment by rwkarg** [comment]
Multiple backend support will address #24

File: controllers/types.go [hunk 1]
```diff
@@ -27,28 +27,59 @@ type NEGStatus struct {
...
 type AutonegStatus struct {
 	AutonegConfig
 	NEGStatus
 }
 
 // Statuses represents the autoneg-relevant structs fetched from annotations
 type Statuses struct {
-	anConfig  AutonegConfig
-	anStatus  AutonegStatus
+	anConfig  OldAutonegConfig
```
Comment (soellman) [inline controllers/types.go:70]: I'd prefer that we also rename the variable names. Maybe we should just drop the "an" and use config and oldConfig?

File: controllers/types.go [hunk 2]
```diff
@@ -27,28 +27,59 @@ type NEGStatus struct {
...
+	nanConfig AutonegConfig
+	anStatus  OldAutonegStatus
+	nanStatus AutonegStatus
 	negStatus NEGStatus
+	negConfig NEGConfig
+	newConfig bool
```
Comment (soellman) [inline controllers/types.go:76]: Maybe flip and note if the config is old.

File: main.go [hunk 3]
```diff
@@ -34,7 +34,7 @@ import (
 	// +kubebuilder:scaffold:imports
 )
 
-const useragent = "google-pso-tool/gke-autoneg-controller/0.9.2-dev"
+const useragent = "google-pso-tool/gke-autoneg-controller/0.9.4-dev"
```
Comment (soellman) [inline main.go:37]: Let's not bump this here - we can do it as part of a release.

**Review by soellman — COMMENTED** [review]
Just a brief review with a few style comments.

My concern here is that we change behavior for existing users' services (with a new status annotation). Can we write out the old status annotation if the user is using the old config annotation?

File: controllers/service_controller.go [hunk 4]
```diff
@@ -27,7 +28,7 @@ import (
 	"sigs.k8s.io/controller-runtime/pkg/reconcile"
 
 	corev1 "k8s.io/api/core/v1"
-	"k8s.io/apimachinery/pkg/api/errors"
+	machineryerrors "k8s.io/apimachinery/pkg/api/errors"
```
[the remark this answers is not in the document]
Comment (rosmo) [reply controllers/service_controller.go]: Changed.

File: controllers/types.go (hunk 1 above)
Comment (rosmo) [reply controllers/types.go]: Dropped an and nan prefixes and went with oldConfig/config/oldStatus/status.

File: main.go (hunk 3 above)
Comment (rosmo) [reply main.go]: Took it out of commit.

File: controllers/types.go (hunk 2 above)
Comment (rosmo) [reply controllers/types.go:76]: Kept it, since it get initialized to false :)

**Comment by rosmo** [comment]
Added the old status writing if the config is old style, this is what an example service annotations look like now (with old config, new config doesn't have any `anthos.cfg.dev` keys):
```
annotations:
    anthos.cft.dev/autoneg: '{"name":"autoneg-test", "max_rate_per_endpoint":1000}'
    anthos.cft.dev/autoneg-status: '{"name":"autoneg-test","max_rate_per_endpoint":1000,"network_endpoint_groups":{"80":"k8s1-e13b8f0c-default-hello-service-80-322e857b"},"zones":["europe-west4-a","europe-west4-b","europe-west4-c"]}'
    cloud.google.com/neg: '{"exposed_ports":{"80":{}}}'
    cloud.google.com/neg-status: '{"network_endpoint_groups":{"80":"k8s1-e13b8f0c-default-hello-service-80-322e857b"},"zones":["europe-west4-a","europe-west4-b","europe-west4-c"]}'
    controller.autoneg.dev/neg-status: '{"backend_services":{"80":{"autoneg-test":{"name":"autoneg-test","max_rate_per_endpoint":1000}}},"network_endpoint_groups":{"80":"k8s1-e13b8f0c-default-hello-service-80-322e857b"},"zones":["europe-west4-a","europe-west4-b","europe-west4-c"]}'
```

**Comment by soellman** [comment]
Thanks Taneli! Let's merge and give it some soak time before we cut a release with this functionality.

**Comment by soellman** [comment]
Fixes #30 
Fixes #24 
Fixes #23 
Fixes #14

**Pull request merged**